vllm/transformers_utils/tokenizer.py

```diff
@@ -67,53 +67,30 @@ def get_tokenizer(
     return tokenizer
 
 
+# Based on huggingface/text-generation-inference/blob/v0.9.4/\
+# server/text_generation_server/models/model.py#L62C9-L62C15
+# under Apache 2.0 license
 def detokenize_incrementally(
     tokenizer: Union[PreTrainedTokenizer, PreTrainedTokenizerFast],
-    prev_output_tokens: List[str],
-    new_token_id: int,
-    skip_special_tokens: bool,
-) -> Tuple[str, str]:
-    """Detokenizes the new token in conjunction with the previous output tokens.
+    all_input_ids: List[int],
+    prefix_offset: int = 0,
+    read_offset: int = 0,
+    skip_special_tokens: bool = False,
+) -> Tuple[str, int, int]:
+    # The prefix text is necessary only to defeat cleanup algorithms in
+    # the decode which decide to add a space or not depending on the
+    # surrounding ids.
+    prefix_text = tokenizer.decode(all_input_ids[prefix_offset:read_offset],
+                                   skip_special_tokens=skip_special_tokens)
+    new_text = tokenizer.decode(all_input_ids[prefix_offset:],
+                                skip_special_tokens=skip_special_tokens)
```

Comment:
Isn't this too expensive? The purpose of the `detokenize_incrementally` function is to avoid de-tokenizing the entire token ids every iteration. Specifically, in the original function, we minimized the overhead of `convert_ids_to_tokens` by caching the previous tokens. In contrast, in the proposed change, the whole token ids are de-tokenized every iteration. Could you measure its impact on the performance?

Reply:
It's not actually the entirety of the token IDs - the offsets are modified every iteration, meaning that there are 5-6 tokens being detokenized at most. I am happy to microbenchmark this, but I wouldn't expect a huge difference.

Reply:
Additionally, now that I think about it, we should be able to leverage rust parallelism from fast tokenizers and perform batch decode faster. I will create a microbenchmark to test the old approach, approach in this PR, and see if batching provides any benefit.